Refactored mafia and mafia_param

aeirya opened the pull request:
did some stuff to increase readability and code quality:

- commented code.
- reduced cognitive complexity of god_page function
- refactored some of the code into functions
- removed duplicate code
- split mafia_param code into multiple files
- game now uses json to store it's configuration and texts
- added some whitespace
- converted player class into a data class

Also added some functionality:

- moved player related code from mafia into its own class
- added team and role classes
(todo:
. use these to refer to roles instead of strings
. add all roles)

sadrasabouri commented:
> did some stuff to increase readability and code quality:
> 
> * commented code.
> * reduced cognitive complexity of god_page function
> * refactored some of the code into functions
> * removed duplicate code
> * split mafia_param code into multiple files
> * game now uses json to store it's configuration and texts
> * added some whitespace
> * converted player class into a data class
> 
> Also added some functionality:
> 
> * moved player related code from mafia into its own class
> * added team and role classes
>   (todo:
>   . use these to refer to roles instead of strings
>   . add all roles)

If you are not going to solve `todo` section in this PR, please issue it.

sadrasabouri commented on Role.py:
```diff
@@ -0,0 +1,107 @@
+    def get_repitition(self) -> int:
+        return int(mafia_params.nRoles[self.name])
+
+    # get desription(self) -> str here
```
Remove these comments and issue them.

sadrasabouri commented on Role.py:
```diff
@@ -0,0 +1,107 @@
+        )
+
+class Roles(enum.Enum):
+    #could've used enum.auto()
```
Remove this too, you can talk about your strategies and decisions in PR comments so that one who want to read about it can find his/her answer easily and keep code simple too.

aeirya replied:
ok .. it was actually a note for myself which I forgot to remove from code

sadrasabouri commented on Role.py:
```diff
@@ -0,0 +1,107 @@
+    return Roles["_".join([word.upper() for word in string.split()])]
+
+roles = dict([(r, r.to_role()) for r in Roles])
+ordered_roles = [roles.get(get(key)) for key in mafia_params.ordered_roles]
```
Last new line is common for code files.
Please consider it in your files.

sadrasabouri commented on mafia.py:
```diff
@@ -1,110 +1,171 @@
 from sys import argv
 from random import randrange, shuffle
+from typing import Callable, Dict, List, Text, Tuple
```
I don't believe `typing` can help us.
Most of these variables are in-app variables and have controlled values so we don't need `typing`.
Instead of that you can make proper docstring for your functions.

aeirya replied:
I personally don't see any reason not to use it.
Sure we could use docstrings too but for more rewarding info imo.
Using too much doc is just gonna get our code cluttered and slow to read by others.
But still, I could remove them after after I'm done with those functions, if u insist.

sadrasabouri replied:
This kind of variable assignment is reminding **_Rust_** for me and I don't want the code be that much frightening to look at it, for instance it may be obvious for reader to understand that `is_...` may return `True` or `False`.

sadrasabouri commented on mafia.py:
```diff
@@ -1,110 +1,171 @@
+        self.n_players = n_players
+        self.roles = roles
+
+    def is_ip_valid(self, ip : str) -> bool:
```
`is_repeated`

aeirya commented:
> did some stuff to increase readability and code quality:
> 
> * commented code.
> * reduced cognitive complexity of god_page function
> * refactored some of the code into functions
> * removed duplicate code
> * split mafia_param code into multiple files
> * game now uses json to store it's configuration and texts
> * added some whitespace
> * converted player class into a data class
> 
> Also added some functionality:
> 
> * moved player related code from mafia into its own class
> * added team and role classes
>   (todo:
>   . use these to refer to roles instead of strings
>   . add all roles)

- added enums for roles, replacing their string representation in the code.
- added a few classes, trying to give the project a better design.
- silenced code factor warnings.

needless to say this issue links directly to #6

sadrasabouri commented on mafia.py:
```diff
@@ -1,110 +1,171 @@
-            elif ip2player[ip].get_state() == "banned":
-                ip2player[ip].set_state("alive")
+def god_page():
+    def kill(player : Player) -> None:
+        # notice banned players cannot be killed
+        if player.is_alive():
+            player.die()
```
add another `if` which makes player `alive` if `kill` is pressed again after player's death.

aeirya replied:
done.
also checks for is_banned now

sadrasabouri commented on mafia.py:
```diff
@@ -1,110 +1,171 @@
+    }
+
+    # filters out invalid requests
+    def get_requests() -> List[Tuple[str, Player]]:
+        requests = []
+        for key in request_mapper.keys():
+            ip = request.args.get(key)
+            
+            if ip is None:
+                continue
+            if not player_manager.is_ip_valid(ip):
+                print("invalid ip / request: ", ip, f"({key})")
+                return -1
+
+            player = player_manager.get_player(ip)
+            requests.append((key, player))
+        return requests
```
I don't get the point of this part please explain it for me and it's use in application

aeirya replied:
it's supposed to check for every possible request type in our request object, and check if the ip address is valid (it was previously done by duplicated lines of ip checking).

what you may find strange at first is that this function returns a list of request types and their arguments.
it isn't uncommon to split a big function to smaller ones with different responsibilities, which I did for the god function.
tho the main reason i made get_requests was that fact i was unsure if it is possible for a request object to have multiple args or not (and i actually guess the answer is no, which means i'll probably clean this part a little bit soon)

sadrasabouri commented on player.py:
```diff
@@ -1,12 +1,42 @@
+    BANNED: str = "banned"
+    all: List[str] = [ALIVE, DEAD, BANNED]
+
+# A player in mafia game
+@dataclass
 class Player:
-    """A player in mafia game"""
-    def __init__(self, ip, username, role, image_name):
-        self.ip = ip
-        self.username = username
-        self.state = "alive"
-        self.role = role
-        self.image_name = image_name
-        self.comment = False
+    ip : str
+    username : str
+    role : str
+    image_name : str
+    state : PlayerState = PlayerState.ALIVE
+    comment : bool = False
```
Change it back.
No need for `typing`.

aeirya replied:
data classes require you to do typing. (it just doesn't work if we don't)
needless to say, it is a much more intuitive way of making classes which are data holder.

sadrasabouri replied:
If you say it, that's OK.

sadrasabouri reviewed (changes requested):
Hi @aeirya 
Thanks for your PR, I reviewed it and there was some points you should consider.
After resolving them I'll check performance (respected to the time of server actions).
Remember to keep it as simple as possible so that anyone can easily read the code and contribute to it.
